[SYCL][DOC] Spec for discard queue events (#4922)
intel/llvm: sycl/doc/extensions/DiscardQueueEvents/SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS.asciidoc

= SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS
:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en

:blank: pass:[ +]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

// This is necessary for asciidoc, but not for asciidoctor
:cpp: C++

== Introduction

IMPORTANT: This specification is a draft.

NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
used by permission by Khronos.

> This document describes an extension that introduces a discard_events property for

Reviewer: ```suggestion
This document describes an extension that introduces a `discard_events` property for

```

SYCL queues.  This property enables developers to inform a SYCL implementation that

> the returned event can be discarded.

Author: Done

== Notice

Copyright (c) 2021 Intel Corporation. All rights reserved.

== Status

Working Draft

This is a preview extension specification, intended to provide early access to
a feature for review and community feedback. When the feature matures, this
specification may be released as a formal extension.

Because the interfaces defined by this specification are not final and are
subject to change they are not intended to be used by shipping software
products.

== Version

> Built On: {docdate} +

Author: Done

Revision: 1

== Contributors

Alexander Flegontov, Intel  +
Greg Lueck, Intel  +
John Pennycook, Intel

== Dependencies

> This extension is written against the SYCL 2020 specification, Revision 3.

Author: Done

== Feature Test Macro

This extension provides a feature-test macro as described in the core SYCL
specification section 6.3.3 "Feature test macros".  Therefore, an
implementation supporting this extension must predefine the macro
`SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` to one of the values defined in the table below.
Applications can test for the existence of this macro to determine if the
implementation supports this feature, or applications can test the macro's
value to determine which of the extension's APIs the implementation supports.

[%header,cols="1,5"]
|===
|Value |Description
|1     |Initial extension version.  Base features are supported.
|===

== Overview

This extension adds `ext::oneapi::property::queue::discard_events` property for `sycl::queue`,

> by using this property the application informs a SYCL implementation that the returned event can be discarded.

Author: Done

In cases where it will be possible, the implementation makes a decision about internal optimization -- avoiding
the creation of low-level event from the backend. The property affects only kernel cases,
i.e. avoid creating low-level event for submission of kernels.

> Performance improvement is the main motivation for using this property.

Author: No, I believe the property should affect only running kernels. Why should we touch other cases? when optimization will be done only for the kernel case. The original purpose of creating this extension is only for kernels.

For example, let's look at the following scenario:
```
USM_pointer = sycl::malloc_device(...);
queue.memcpy(USM_pointer, , ).wait(); // The user will have an exception with errc::invalid on event.wait(), 
queue.submit(...);
queue.wait();
```

Why do we need to restrict the user and force him to rewrite the code to use `q.wait()` after `queue.memcpy()`?
Is there a good reason for this restriction?

> Performance improvement is the main motivation for using this property.

Reviewer 2: Isn't the optimization equally useful for the memory copy operations?  Why wouldn't we apply the optimization also in that case?

It seems like we expect people to use this option also with `in_order`.  In that case, there would be no need to call `q.wait()` after `queue.memcpy()`.  You could instead just immediately schedule the next command.

> Performance improvement is the main motivation for using this property.

Author: I checked what you suggested about other queue operations and I agree with you that this optimization is also useful for other queue operations, and we can do this extension for any of them. 
Aligned the doc according to that.

Below is a usage example:
[source,c++]
----
  sycl::property_list props{ext::oneapi::property::queue::discard_events{}
                            property::queue::in_order{}};
  sycl::queue Queue( props );

  // some USM preparations ..

  sycl::event e1, e2, e3;

  // returning "invalid" events from each submission function:
  e1 = Queue.parallel_for(NDRange, [=](nd_item<1> item){ do_smth1(); });

  e2 = Queue.single_task([=](){ do_smth2(); });

  e3 = Queue.submit([&](handler &CGH) { CGH.parallel_for(NDRange, [=](nd_item<1> item){ do_smth3(); }); });

  Queue.wait();
----

In the example above, the application doesn't use sycl events: `e1`, `e2`, `e3`
and is waiting for the end of work by `queue::wait()`. The returned events will be
_invalid_ events. _invalid_ event is a `sycl::event` with a status `invalid` and the event returns from each kernel

> submission when queue is created with the `discard_events` property.

Author: Done.

See the description of behavior for this event below for details.

> Here, only those methods for the _invalid_ event are described that have behavior different from the default event behavior:

Reviewer 2: ```suggestion
Here, only those member functions for the _invalid_ event are described that have behavior different from the default event behavior:

```

Author: Done.

[source,c++]
----
// must throw an exception with the errc::invalid error code.
void wait();

// if invalid event is passed into the function, must throw an exception with the errc::invalid error code.
static void wait(const std::vector<event> &eventList);

// must throw an exception with the errc::invalid error code.
void wait_and_throw();

// if invalid event is passed into the function, must throw an exception with the errc::invalid error code.
static void wait_and_throw(const std::vector<event> &eventList);

// must return info::event_command_status::ext_oneapi_unknown
get_info<info::event::command_execution_status>() const;

> ----

Reviewer 3: Should we say anything special for `get_wait_list()`? I believe it can follow the original semantic.

> ----

Reviewer 2: Are you suggesting that `get_wait_list()` should return normal (i.e. not "invalid") events?

Thinking about this more, it would probably be easier to just say that `get_wait_list()` throws `errc::invalid`.

> ----

Author: BTW, the default event returns a zero-size vector for `get_wait_list()`, So I believe the invalid event also can save the semantic, Should we change it to throwing `errc::invalid` just in case?

> ----

Reviewer 2: Returning an empty set of events for the default-constructed event make sense because that event is always "complete".  Thus, there are never any other events to wait for.

I do not think it makes sense to return an empty set of events for an "invalid" event.  If we did, code like this would silently do the wrong thing:

```
cgh.depends_on(myevt.get_wait_list());
```

Whatever we do, code like I show above should cause an exception.  This would happen if `get_wait_list()` returns a vector of "invalid" events, or it would happen if `get_wait_list()` itself throws an exception.  Throwing an exception from `get_wait_list()` is easier, so that is probably what makes sense.

> ----

Reviewer 3: ```
cgh.depends_on(myevt.get_wait_list());
```
Interesting example, I was think about such a case:

```
sycl::queue QReg{...}; // regular queue
sycl::queue QDiscardEvents{...}; // a queue with ext::oneapi::property::queue::discard_events{}
auto RegEvent = QReg.submit(...);
auto DiscardedEvent = QDiscardEvents.submit(... CGH.depends_on(RegEvent); ...); // A commands depends on a regular event from another queue

auto WaitList = DiscardedEvent.get_wait_list(); // get_wait_list line
```

I would not expect that `get_wait_list line` returns zero event list(in general case) or list of invalid events.
So, throwing an exception from `get_wait_list` sounds reasonable.

The behavior when _invalid_ event is passed into handler API:
[source,c++]
----
// must throw an exception with the errc::invalid error code.
handler::depends_on(event Event)

// must throw an exception with the errc::invalid error code.
handler::depends_on(const std::vector<event> &Events)
----

> 

Author: Done.

== DPCPP implementation limitations

The `discard_events` property should be taken as a hint for SYCL implementation.
In all kernel cases, SYCL implementation returns _invalid_ event, regardless of whether the optimization was done.

> Optimization occurs if the queue has `discard_events` property and all features presented below are not met:

Reviewer 2: I think the title of this section could be improved.  The wording "implementation limitations" sounds too negative.  Maybe "Optimization behavior for DPC++"?  We should also make it clear that this is a non-normative part of the spec.  I'd suggest wording like this for the initial paragraph:

> This non-normative section describes the conditions when the DPC++ implementation provides an optimization benefit for the `discard_events` property.
>
> * The queue must be constructed with the `in_order` property.
> * A kernel submitted to the queue must not use the fallback assert feature.
> * A kernel submitted to the queue must not use streams or buffer /  image accessors.  However, local accessors do not inhibit optimization.

Author: I like "Optimization behavior for DPC++", applied your suggestion.

specification section 6.3.3 "Feature test macros".  Therefore, an
implementation supporting this extension must predefine the macro
`SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` to one of the values defined in the table below.
Applications can test for the existence of this macro to determine if the
implementation supports this feature, or applications can test the macro's
value to determine which of the extension's APIs the implementation supports.

[%header,cols="1,5"]
|===
|Value |Description
|1     |Initial extension version.  Base features are supported.
|===

== Overview

This extension adds `ext::oneapi::property::queue::discard_events` property for `sycl::queue`,
by using this property the application informs a SYCL implementation that the returned event can be discarded.

In cases where it will be possible, the implementation makes a decision about internal optimization -- avoiding
the creation of low-level event from the backend. The property affects only kernel cases,
i.e. avoid creating low-level event for submission of kernels.
Performance improvement is the main motivation for using this property.

Below is a usage example:
[source,c++]
----

>   sycl::property_list props{ext::oneapi::property::queue::discard_events{}

Reviewer 3: ```suggestion
  sycl::property_list props{ext::oneapi::property::queue::discard_events{}, 

```

Author: Done.

 - A queue operation submitted to the queue must not use streams or buffer / image accessors. However, local accessors do not inhibit optimization.
 - Any queue operations using Level Zero backend temporarily work without optimization.

*The benefit is that a low-level event is not created from backend, thereby saving time.

See the behavior details for each condition below:

=== Using out-of-order queue

No optimization if a queue is created with the `discard_events` property and
the property list does not include `in_order` property.

=== Using fallback assert feature

No optimization if the application calls the `assert` macro from a command that is submitted to the queue unless
the device has native support for assertions (as specified by `aspect::ext_oneapi_native_assert`).

=== Using streams or buffer / image accessors (excluding local accessors)

No optimization if a queue operation that uses stream objects or buffer / image accessors is submitted to a queue created with
the `discard_events` property. But using local accessors does not affect optimization.

=== Using Level Zero backend

Since Level Zero Plugin support is required to be able to not create a low-level event,

> so any queue operations using the Level Zero backend temporarily work without optimization.

Reviewer: ```suggestion
any queue operations using the Level Zero backend temporarily work without optimization.

```

a feature for review and community feedback. When the feature matures, this
specification may be released as a formal extension.

Because the interfaces defined by this specification are not final and are
subject to change they are not intended to be used by shipping software
products.

== Version

Built On: {docdate} +
Revision: 1

== Contributors

Alexander Flegontov, Intel  +
Greg Lueck, Intel  +
John Pennycook, Intel

== Dependencies

This extension is written against the SYCL 2020 specification, Revision 3.

== Feature Test Macro

This extension provides a feature-test macro as described in the core SYCL
specification section 6.3.3 "Feature test macros".  Therefore, an
implementation supporting this extension must predefine the macro
`SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` to one of the values defined in the table below.
Applications can test for the existence of this macro to determine if the
implementation supports this feature, or applications can test the macro's
value to determine which of the extension's APIs the implementation supports.

[%header,cols="1,5"]
|===
|Value |Description
|1     |Initial extension version.  Base features are supported.
|===

== Overview

> This extension adds `ext::oneapi::property::queue::discard_queue_events` property for `sycl::queue`,

Author: ok, I will change the name to `ext::oneapi::property::queue::discard_events`, but the macro name for feature-test should remain unchanged, i.e. `SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS`, right?

> This extension adds `ext::oneapi::property::queue::discard_queue_events` property for `sycl::queue`,

Reviewer 2: Correct.

_invalid_ event API behavior:
[source,c++]
----
// the behavior is the same as for the default event, i.e. must throw an exception with the errc::invalid_object error code
cl_event get() const

// the behavior is the same as for the default event, i.e. must return true
bool is_host() const;

// the behavior is the same as for the default event, i.e. must return backend::host
backend get_backend() const noexcept;

// the behavior is the same as for the default event, i.e. must return a zero-sized vector
std::vector<event> get_wait_list();

// must throw an exception with the errc::invalid_object error code.
void wait();

// passing invalid event in the function must throw an exception with the errc::invalid_object error code.
static void wait(const std::vector<event> &eventList);

// must throw an exception with the errc::invalid_object error code.
void wait_and_throw();

// passing invalid event in the function must throw an exception with the errc::invalid_object error code.
static void wait_and_throw(const std::vector<event> &eventList);

// returns info::event_command_status::ext_oneapi_unknown status
get_info<info::event::command_execution_status>() const;

> // the behavior is the same as for the default event, i.e. returns 0

Reviewer 3: Hm, does default event really returns `0`? Is it defined by the spec? If no suggest removing `i.e. returns 0`.

Author: Yes, the default event returns 0, but I cannot find the exact statement of that in spec. So I will remove it. Thanks.

get_info<info::event::reference_count>() const;

// must throw an exception with the errc::invalid_object error code.
template <info::event_profiling param>
typename param::return_type get_profiling_info() const;
----

The behavior when passing _invalid_ event to handler API:
[source,c++]
----
// must throw an exception with the errc::invalid_object error code.
handler::depends_on(event Event)

// must throw an exception with the errc::invalid_object error code.
handler::depends_on(const std::vector<event> &Events)
----

> == Limitations

Reviewer 3: ```suggestion
== DPCPP implementation limitations
```
This is planned limitations for the first version of implementation which we may lift in future, right?
If so, maybe rename it to something like this?

Author: Right, I will update.

The `discard_queue_events` property should be taken as a hint for SYCL implementation.
In all kernel cases, SYCL implementation returns _invalid_ event, regardless of whether the optimization was done.
Optimization occurs if the queue has `discard_queue_events` property and all features presented below are not met:

 - Using out-of-order queue or using together with `enable_profiling` property
 - Using https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Assert/SYCL_ONEAPI_ASSERT.asciidoc[fallback assert feature]
 - Using streams, buffer/image accessors (excluding local accessors)

> 

Reviewer 3: Shouldn't we mention reductions here?

Author: I'm not sure if this needs to be written explicitly, because that case is already covered by using accessors. If you use accessors for reduction then it is already limited.

Reviewer 3: I thought we support USM based reductions as well.
Tagging @v-klochkov

Author: > I thought we support USM based reductions as well.

yes, It is correct!

Accessor based reduction is NOT supported.
USM based reductions is supported.

Reviewer: Doesn't the USM-based reduction use events behind the scenes, though?  A reduction might be enqueued as multiple kernels, and I believe that events/accessors are used as an implementation detail to keep those kernels ordered.  I don't think there's a special-case for reductions submitted to in-order queues.  @v-klochkov?

Reviewer 4: If I understood it correctly, call of event::wait would cause exception, but reduction does not call wait().
Reduction implementation may create more than one kernel, and it connects them using either accessors (which has read after write sync builtin), or using depends_on(), which should work correctly with the new proposed attribute (e.g.: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/reduction.hpp#L1953)

IMO, reduction should keep working for both USM and buffer/accessors.

See details for each below:

=== Using out-of-order queue and/or using together with enable_profiling property

No optimization if a queue is created with the `discard_queue_events` property and
the property list includes the `enable_profiling` property or does not include `in_order` property.

=== Using fallback assert feature

If a kernel that uses the fallback assert feature is submitted to a queue created with

> the `discard_queue_events` property, the implementation must throw an exception with the errc::invalid_object error code.

Reviewer 3: Could you please tell why this is a "hard" limitation(throw an exception) rather than a "soft"(optimization doesn't work in this case) one?

Author: This is due to internal SYCL implementation. Since in that case, we need to somehow pass the information to the kernel launch function that the kernel uses the `assert` macro or not. I checked the code and found that if that limitation will be"soft" then we will have additional overhead(twice capture/release mutex `GlobalHandler::instance().getHandlerExtendedMembersMutex()`). --- this is due to a violation of our ABI policy as I cannot add a new field for CG and handler class.

But if we throw an exception as soon as we know that the `fallback assert feature` is enabled then we will not do this extra code for the case when `fallback assert feature` is disabled.

Moreover, fallback assert is not about performance, it has overhead even if the `assert` statement is not written in the kernel. And since the user who will use the new `SYCL_EXT_ONEAPI_DISCARD_QUEUE_EVENTS` extension clearly cares about performance, we can do it for fewer cases, but do it better/faster.

Author: Well, I looked at it again and found that we can make it "soft" and it will be inexpensive. I changed the doc.

It is important to disable the logic for detecting assertion failures in kernels in principle
since it is not enough just not to use the `assert` macro inside kernel code or compile with `NDEBUG`.
To disable the logic, developer must compile with SYCL_DISABLE_FALLBACK_ASSERT macro defined,
for more details see https://github.com/intel/llvm/blob/sycl/sycl/doc/PreprocessorMacros.md[document]

[source,c++]
----
  sycl::property_list props{ext::oneapi::property::queue::discard_queue_events{},
                            property::queue::in_order()};
  sycl::queue Queue( props );
  // if the submission of kernel1 is compiled with the logic for detecting assertion failures enabled,
  // then the submission of the kernel throws an exception with the errc::invalid_object error code.
  Queue.parallel_for<class kernel1>(NDRange, [=](nd_item<1> item){ do_smth(); });
----

=== Using streams, buffer/image accessors (excluding local accessors)

No optimization if a kernel that uses stream objects, buffer or image accessors is submitted to a queue created with
the `discard_queue_events` property. But using local accessors does not affect optimization.

> *NOTE : This extension is only compatible with kernels using Unified Shared Memory (USM) pointers.

Reviewer 3: Not sure that with latest updates it's still true.

Author: ok, removed the note
